File: Day7/maximum_number_of_events_that_can_be_attended.cc
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

class Solution {
public:
    int maxEvents(vector<vector<int>>& events) {
        sort(events.begin(), events.end());
        priority_queue<int, vector<int>, greater<int>> pq;

        int j = 0, maxEvents = 0;
        int n = events.size();
        int maxDay = 0;

        for (const auto& event : events) {
            maxDay = max(maxDay, event[1]);
        }

        for (int day = 1; day <= maxDay; ++day) {
            while (j < n && events[j][0] <= day) {
                pq.push(events[j][1]);
                j++;
            }

            while (!pq.empty() && pq.top() < day) {
                pq.pop();
            }

            if (!pq.empty()) {
                pq.pop();
                maxEvents++;
            }
        }

        return maxEvents;
    }
};

File: Day7/maximum_number_of_events_that_can_be_attended.java
import java.util.*;

class Solution {
    public int maxEvents(int[][] events) {
        Arrays.sort(events, Comparator.comparingInt(a -> a[0]));
        PriorityQueue<Integer> pq = new PriorityQueue<>();

        int j = 0, maxEvents = 0;
        int n = events.length;
        int maxDay = 0;

        for (int[] event : events) {
            maxDay = Math.max(maxDay, event[1]);
        }

        for (int day = 1; day <= maxDay; day++) {
            while (j < n && events[j][0] <= day) {
                pq.offer(events[j][1]);
                j++;
            }

            while (!pq.isEmpty() && pq.peek() < day) {
                pq.poll();
            }

            if (!pq.isEmpty()) {
                pq.poll();
                maxEvents++;
            }
        }

        return maxEvents;
    }
}
